[WIP] Refactor dynamic tuning and add new global option

TomasKorbar opened the pull request:
See commit messages for description of changes.

TomasKorbar committed:
Merge option bool and get_bool methods

New get_bool method is in ParserOptions class

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar committed:
Rework Tuned's way of parsing options

ParserOptions will contain the most used methods for parsing of
plugin tuning options

ParserInstanceOptions will be used specifically for parsing of
plugin instance options got from profile and dividing them
into plugin tuning options, which serve for technical parameters
of tuning, and instance management options, which are used for
changing of instance behaviour (for now only dynamic tuning toogle)

Add dynamic_tuning_enabled_by_default method to Plugin class.
By overriding this method plugin can say that it wants to run
its dynamic tuning by default without need of explicit permission
in profile.

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar committed:
Refactor Instance class and change its use in plugins

Instance class now uses its own parser (ParserInstanceOptions)

_has_dynamic_tuning has been changed to dynamic_tuning_enabled
plugins can use method forbid_dynamic_tuning if it for some
reason does not want to run its dynamic tuning otherwise they
are not allowed to change its value

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

olysonek commented on tuned/plugins/base.py:
```diff
@@ -41,6 +41,24 @@ def __init__(self, monitors_repository, storage_factory, hardware_inventory, dev
+	def is_static_tuning_supported(cls):
+		return True
+
+	# If plugin wants to run it's dynamic tuning by default then it must
```
it's -> its

TomasKorbar replied:
Fixed in https://github.com/olysonek/tuned/pull/1/commits/8e2d179fbb93378e78d5c6048759b2e6546ac6cc

olysonek commented on tuned/plugins/base.py:
```diff
@@ -120,6 +138,26 @@ def initialize_instance(self, instance):
+		# initialize values required for dynamic tuning only if plugin is
+		# capable of dynamic tuning and instance_init allows it but if it is
+		# capable only of dynamic tuning then initialize them by default
+		plugins_dynamic_tuning_enabled = self._global_cfg.get_bool(
```
"plugins" -> "plugin" in "plugins_dynamic_tuning_enabled" would make more sense, no?

TomasKorbar replied:
Yes. Fixed in https://github.com/olysonek/tuned/pull/1/commits/560ffba2667702f7ee714462f222a5aace9fbdac

olysonek commented on tuned/plugins/base.py:
```diff
@@ -120,6 +138,26 @@ def initialize_instance(self, instance):
 		log.debug("initializing instance %s (%s)" % (instance.name, self.name))
 		self._instance_init(instance)
+		# initialize values required for dynamic tuning only if plugin is
+		# capable of dynamic tuning and instance_init allows it but if it is
```
`instance_init` -> `_instance_init`

TomasKorbar replied:
Fixed in https://github.com/olysonek/tuned/pull/1/commits/8e2d179fbb93378e78d5c6048759b2e6546ac6cc

olysonek commented on tuned/plugins/base.py:
```diff
@@ -41,6 +41,24 @@ def __init__(self, monitors_repository, storage_factory, hardware_inventory, dev
+		return False
+
+	@classmethod
+	def is_static_tuning_supported(cls):
```
The method naming is inconsistent: `is_static_tuning_supported` vs. `dynamic_tuning_enabled_by_default`. I would drop the `is_` prefix in the two methods above.

TomasKorbar replied:
Fixed in https://github.com/olysonek/tuned/pull/1/commits/ba9aa4df75ee9e5b9d0976dadc07bb5103bf84d0

olysonek commented on tuned/plugins/base.py:
```diff
@@ -120,6 +138,26 @@ def initialize_instance(self, instance):
+			instance._has_dynamic_tuning = False
+			return
+		dynamic_init_by_choice = (self._option_bool(instance.options.get("dynamic", False)) and
+								  self.is_dynamic_tuning_supported())
```
In another commit, log an error if `options["dynamic"] is True` but dynamic tuning is not supported.

Also, shouldn't "dynamic" be popped from the `options` dictionary?

TomasKorbar replied:
I changed this in commit https://github.com/olysonek/tuned/pull/1/commits/804434c4e7d9f3e81936a9c3e7934cbb97d2899f

TomasKorbar replied:
And added log message in new commit https://github.com/olysonek/tuned/pull/1/commits/3e86c691648317c2e3e1d8ccdda6c818068b8704

olysonek commented on tuned/plugins/base.py:
```diff
@@ -41,6 +41,24 @@ def __init__(self, monitors_repository, storage_factory, hardware_inventory, dev
+	# If plugin wants to run it's dynamic tuning by default then it must
+	# override this
+	@classmethod
+	def dynamic_tuning_enabled_by_default(cls):
```
This method is not used anywhere in this commit, so it should get added in a separate commit.

olysonek commented on tuned/plugins/plugin_cpu.py:
```diff
@@ -37,6 +37,14 @@ def __init__(self, *args, **kwargs):
+		return True
+
+	@classmethod
+	def dynamic_tuning_enabled_by_default(cls):
```
This method is not used anywhere in this commit, so it should get added in a separate commit.

olysonek commented on tuned/plugins/plugin_disk.py:
```diff
@@ -24,6 +24,14 @@ def __init__(self, *args, **kwargs):
+		return True
+
+	@classmethod
+	def dynamic_tuning_enabled_by_default(cls):
```
This method is not used anywhere in this commit, so it should get added in a separate commit.

olysonek commented on tuned/plugins/plugin_eeepc_she.py:
```diff
@@ -20,6 +20,18 @@ def __init__(self, *args, **kwargs):
+		return False
+
+	@classmethod
+	def dynamic_tuning_enabled_by_default(cls):
```
This method is not used anywhere in this commit, so it should get added in a separate commit.

olysonek commented on tuned/plugins/plugin_net.py:
```diff
@@ -21,6 +21,14 @@ def __init__(self, *args, **kwargs):
+		return True
+
+	@classmethod
+	def dynamic_tuning_enabled_by_default(cls):
```
This method is not used anywhere in this commit, so it should get added in a separate commit, or in the commit where it's actually used.

olysonek commented on tuned/plugins/base.py:
```diff
@@ -120,6 +138,26 @@ def initialize_instance(self, instance):
+		dynamic_init_by_default = (self.is_dynamic_tuning_supported() and
+								   not self.is_static_tuning_supported())
+		if dynamic_init_by_choice or dynamic_init_by_default:
+			self._instance_init_dynamic_tuning(instance)
```
This logic will enable dynamic tuning if `dynamic_init_by_default is True`, even though the user set `dynamic=False` in the profile. You could argue that such configuration doesn't make sense for a plugin that only supports dynamic tuning, however Tuned needs to behave reasonably even in such a case. Behave reasonably == should disable dynamic tuning.

TomasKorbar replied:
This has been also changed and in my opinion solved in https://github.com/olysonek/tuned/pull/1/commits/804434c4e7d9f3e81936a9c3e7934cbb97d2899f

olysonek commented on tuned/plugins/repository.py:
```diff
@@ -32,6 +32,15 @@ def _set_loader_parameters(self):
+		only_dynamic = (plugin_cls.is_dynamic_tuning_supported() and
+						not plugin_cls.is_static_tuning_supported())
+		if not plugins_dynamic_tuning_enabled and only_dynamic:
+			log.info("plugin %s supports only dynamic tuning and that is disabled in global config, stopping his initialization" % (plugin_cls.name))
```
his initialization -> its initialization

TomasKorbar replied:
Fixed in https://github.com/olysonek/tuned/pull/1/commits/8fa2898f55c2c79d402c8f13b1bcd76b4b48456f

olysonek commented on tuned/plugins/repository.py:
```diff
@@ -32,6 +32,15 @@ def _set_loader_parameters(self):
 	def create(self, plugin_name):
 		log.debug("creating plugin %s" % plugin_name)
 		plugin_cls = self.load_plugin(plugin_name)
+		plugins_dynamic_tuning_enabled = self._global_cfg.get_bool(
```
"plugins" -> "plugin" in "plugins_dynamic_tuning_enabled" would make more sense, no?

TomasKorbar replied:
Yes. Fixed in https://github.com/olysonek/tuned/pull/1/commits/560ffba2667702f7ee714462f222a5aace9fbdac

olysonek commented on tuned/plugins/repository.py:
```diff
@@ -32,6 +32,15 @@ def _set_loader_parameters(self):
+						not plugin_cls.is_static_tuning_supported())
+		if not plugins_dynamic_tuning_enabled and only_dynamic:
+			log.info("plugin %s supports only dynamic tuning and that is disabled in global config, stopping his initialization" % (plugin_cls.name))
+			return None
```
I think the whole change of this file could be in a separate commit.

olysonek reviewed (changes requested):
This is an incomplete review, there may be some non-sensical comments :). Sorry, I haven't had time yet to finish it.

TomasKorbar committed:
Add log message to initialize_instance method in Plugin class

If plugins dynamic tuning is permitted in profile but it does not
support dynamic tuning then log error about it.

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar committed:
Stop initializing plugins supporting only dynamic tuning if it is disabled

Plugins like eeepce_she, which supports only dynamic tuning will not be
initialized if their dynamic tuning is forbidden in global configuration.

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar committed:
Refactor plugin init of variables needed for dynamic tuning

From now each plugins class has class methods dynamic_tuning_supported
and static_tuning_supported, which serves as a confirmation about
what is plugin capable of and if it is needed to initialize
variables for dynamic tuning (monitors etc.)

From now on it is possible to add for example
'plugin_disk_dynamic_tuning=0'
to tuned's global configuration file and this will force dynamic
tuning of all instances of disk plugin to disable

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar commented:
Squashed commits addressing problems and changed history according to your proposals.

olysonek commented on tuned/plugins/base.py:
```diff
@@ -120,6 +130,26 @@ def initialize_instance(self, instance):
+		# capable of dynamic tuning and _instance_init allows it but if it is
+		# capable only of dynamic tuning then initialize them by default
+		plugin_dynamic_tuning_enabled = self._global_cfg.get_bool(
+			"plugin_"+ self.name + "_dynamic_tuning",
```
How about the following name for the option? I think it would be more understandable:
`plugin_"+ self.name + "_dynamic_tuning_allowed"`

TomasKorbar replied:
Yeah it will provide more clarity. Changed.

TomasKorbar committed:
Remove tests from PluginBaseTestCase

I moved some responsibilites from Plugin class so these tests will
now be moved into the their right testcases.

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar committed:
Remove test from CommandsTestCase

I moved get_bool method from Commands class so this test will need
to be moved too.

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar committed:
Add ParserOptionsTestCase

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

TomasKorbar committed:
Add ParserInstanceOptionsTestCase

Signed-off-by: Tomas Korbar <tkorbar@redhat.com>

olysonek commented on .travis.yml:
```diff
@@ -4,15 +4,10 @@ services:
-    - package_manager=dnf
- - name: "fedora 28 python 3 test suite"
+ - name: "fedora 30 python 3 test suite"
    env:
```
I think this commit can be dropped as we no longer use Travis.

TomasKorbar replied:
Yeah. Dropped

TomasKorbar commented:
Updated tests.

olysonek commented on tuned/utils/parsers/parser_instance_options.py:
```diff
@@ -0,0 +1,47 @@
+			tuple (tuning_options, instance_management_options)
+		"""
+		# TODO: _has_dynamic_options is a hack
+		plug = self._plugin
```
This seems like unnecessary obfuscation. Just use `self._plugin` everywhere.

olysonek commented on tuned/utils/parsers/parser_instance_options.py:
```diff
@@ -0,0 +1,47 @@
+		tuning_effective = plug._get_config_options().copy()
+		instance_effective = self._instance_management_options()
+		for key in options:
+			if key in instance_effective.keys():
```
I think `.keys()` is unnecessary.

olysonek commented on tuned/utils/parsers/parser_instance_options.py:
```diff
@@ -0,0 +1,47 @@
+		self._plugin = plugin
+
+	def _instance_management_options(self):
+		return {"dynamic": self._plugin.dynamic_tuning_enabled_by_default()}
```
My feeling is that this class should not be concerned with defaults and that this should be handled somewhere else. But we can change that later.

olysonek commented on tuned/utils/parsers/parser_options.py:
```diff
@@ -0,0 +1,15 @@
+class ParserOptions():
```
Inherit from `object`, until we make an official decision to drop python2 support.

olysonek commented on tuned/plugins/base.py:
```diff
@@ -135,8 +135,16 @@ def initialize_instance(self, instance):
+			dynamic_init_by_choice = False
+			log.error("Plugin %s does not support dynamic tuning" % (self.name))
+		else:
+			dynamic_init_by_choice = False
```
Uff, that's long. I think that just adding the following lines would suffice, no?
```
if (instance.dynamic_tuning_enabled and not self.dynamic_tuning_supported()):
    log.error("Plugin %s does not support dynamic tuning" % (self.name))
```

olysonek commented on tuned/plugins/base.py:
```diff
@@ -93,17 +93,6 @@ def _get_config_options_used_by_dynamic(self):
 		"""List of config options used by dynamic tuning. Their previous values will be automatically saved and restored."""
 		return []
 
-	def _get_effective_options(self, options):
```
You've left behind `_get_effective_options` in the bootloader plugin.

olysonek commented on tests/plugins/test_base.py:
```diff
@@ -42,17 +42,6 @@ def setUp(self):
 			profiles.variables.Variables())
 
-	def test_get_effective_options(self):
-		self.assertEqual(self._plugin._get_effective_options(\
```
This change should happen in the commit where you remove `_get_effective_options`.

olysonek commented on tests/plugins/test_base.py:
```diff
@@ -42,17 +42,6 @@ def setUp(self):
-			'default_option2': 'default_value2'})
-
-	def test_option_bool(self):
-		self.assertTrue(self._plugin._option_bool(True))
```
This change should happen in the commit where you remove `_option_bool`.

olysonek commented on tests/utils/test_commands.py:
```diff
@@ -20,18 +20,6 @@ def setUp(self):
 
-	def test_get_bool(self):
-		positive_values = ['y','yes','t','true']
-		negative_values = ['n','no','f','false']
```
This change should happen in the commit where `commands.get_bool` is removed.

olysonek commented on tests/utils/parsers/test_parser_options.py:
```diff
@@ -0,0 +1,20 @@
+from tuned.utils.parsers.parser_options import ParserOptions
+
+class ParserOptionsTestCase(unittest2.TestCase):
+
```
The test case should be added in a single commit along with the ParserOptions class.

olysonek commented on tests/utils/parsers/test_parser_instance_options.py:
```diff
@@ -0,0 +1,32 @@
+from tuned.utils.parsers.parser_instance_options import ParserInstanceOptions
+from tuned.plugins.base import Plugin
+
+class ParserInstanceOptionsTestCase(unittest2.TestCase):
```
Add the test case in the same commit where you add the ParserInstanceOptions class.

olysonek commented on tuned/utils/parsers/parser_options.py:
```diff
@@ -3,13 +3,14 @@ class ParserOptions():
 		pass
 
-	def option_bool(self, value):
+	def get_bool(self, value):
```
Why didn't you name it `get_bool` in the first place? :)

olysonek commented:
Ehm.. my master branch was a bit outdated, which explains why there were no merge conflicts :). I fixed that now... I'd appreciate it if you could rebase.
